resolwe_bio/tools/expression_aggregator.py

```diff
@@ -0,0 +1,155 @@
+    return parser.parse_args()
+
+
+def load_expression(fn=None, sep='\t'):
+    """Read expressions from file."""
+    with gzip.open(fn, 'rt') as f:
+        reader = csv.DictReader(f, delimiter=sep)
+        return {row['Gene']: float(row['Expression']) for row in reader}
+
+
+def get_indices(descriptors, descriptor):
+    """Return positions of my_element in my_list."""
+    return {i for i, x in enumerate(descriptors) if x == descriptor}
+
+
+def get_values(expressions, descriptors, gene, descriptor):
+    """Return expressions of a gene with matching descriptor."""
+    indices = get_indices(descriptors, descriptor)
+    return [expression[gene] for i, expression in enumerate(expressions) if i in indices and gene in expression]
+
+
+def load_expressions(aggregator, expression_fns=[], sep='\t', descriptors=[]):
+    """Read expressions from files."""
+    raw_expressions = [load_expression(expression_fn, sep) for expression_fn in expression_fns]
+    if aggregator:
+        raw_expressions.extend(aggregator['raw_expressions'])
+        descriptors.extend(aggregator['descriptors'])
+    genes = {key for raw_expression in raw_expressions for key in raw_expression.keys()}
+    return raw_expressions, descriptors, {
+        gene: {
+            descriptor: get_values(raw_expressions, descriptors, gene, descriptor)
+            for descriptor in set(descriptors)
+            if get_values(raw_expressions, descriptors, gene, descriptor)
+        }
+        for gene in genes
+    }
+
+
+def generate_statistic(expression, gene, attribute, expression_type):
+    """Get box plot statistic for expressions of a single gene and attribute."""
+    min_val = min(expression)
+    max_val = max(expression)
+    median = np.percentile(expression, 50.0)
+    q1 = np.percentile(expression, 25.0)
+    q3 = np.percentile(expression, 75.0)
+    iqr = q3 - q1
+    lowerwhisker = max(min_val, q1 - 1.5 * iqr)
+    upperwhisker = min(max_val, q3 + 1.5 * iqr)
+    return {
+        'attribute': attribute,
+        'gene': gene,
+        'exp_types': [expression_type],
+        'min': min_val,
+        'max': max_val,
+        'median': median,
+        'q1': q1,
+        'q3': q3,
+        'lowerwhisker': lowerwhisker,
+        'upperwhisker': upperwhisker
+    }
+
+
+def get_statistics(expressions, expression_type):
+    """Get box plot statistics for expressions of all genes and attributes."""
+    return {
+        gene: [generate_statistic(exp, gene, descriptor, expression_type) for descriptor, exp in expression.items()]
+        for gene, expression in expressions.items()
+    }
+
+
+def output_json(statistics, fname=None, compressed=False):
+    """Write json to file."""
+    if not compressed:
+        with open(fname, 'w') as f:
+            json.dump(statistics, f)
+    else:
+        with gzip.open(fname, 'wt') as f:
+            json.dump(statistics, f)
+
+
+def load_json(fname, compressed=False):
+    """Read json from file."""
+    if not compressed:
```

Comment:
load_fn = gzip.open if compressed else json.load
    with open(fname, 'r') as f:
        return load_fn(f)

Reply:
This fails with
```
Traceback (most recent call last):
  File "/usr/local/bin/resolwe/1/expression_aggregator.py", line 161, in <module>
    main()
  File "/usr/local/bin/resolwe/1/expression_aggregator.py", line 146, in main
    aggregator = load_json(args.aggregator, True)
  File "/usr/local/bin/resolwe/1/expression_aggregator.py", line 117, in load_json
    return load_fn(f)
  File "/usr/lib/python2.7/gzip.py", line 34, in open
    return GzipFile(filename, mode, compresslevel)
  File "/usr/lib/python2.7/gzip.py", line 94, in __init__
    fileobj = self.myfileobj = __builtin__.open(filename, mode or 'rb')
TypeError: coercing to Unicode: need string or buffer, file found
```